src/main/java/com/emc/pravega/perf/PravegaPerfTest.java

```diff
@@ -259,131 +314,278 @@ private static void parseCmdLine(String[] args) {
                     String payload = String.format("%-" + messageSize + "s", val);
+                    String key;
+                    if (isRandomKey) {
+                        key = Integer.toString(producerId + new Random().nextInt());
+                    } else {
+                        key = Integer.toString(producerId);
+                    }
+                   
                     // event ingestion
-                    long now = System.currentTimeMillis();
-                    retFuture = produceStats.runAndRecordTime(() -> {
-                                String key;
-                                if (isRandomKey) {
-                                    key = Integer.toString(producerId + new Random().nextInt());
-                                } else {
-                                    key = Integer.toString(producerId);
-                                }
+                    retFuture = stats.writeAndRecordTime(() -> {
                                 return fn.apply(key, payload);
                             },
-                            now,
-                            payload.length(), executor);
-                    //If it is a blocking call, wait for the ack
-                    if ( blocking ) {
-                        try {
-                            retFuture.get();
-                        } catch (InterruptedException  | ExecutionException e) {
-                            e.printStackTrace();
-                        }
-                    }
+                            payload.length());
 
                 }
+
                 long timeSpent = System.currentTimeMillis() - loopStartTime;
-                // wait for next event
-                try {
-                    //There is no need for sleep for blocking calls.
-                    if ( !blocking ) {
-                        if ( timeSpent < 1000) {
-                            Thread.sleep(1000 - timeSpent);
-                        }
-                    }
-                } catch (InterruptedException e) {
-                    // log exception
-                    System.exit(1);
-                }
-            }
+                     if (timeSpent < 1000) {
+                          Thread.sleep(1000 - timeSpent);
+                     }
+
+                diffTime = System.currentTimeMillis() - StartTime; 
+ 
+            } while(diffTime < mSeconds);
+
             producer.flush();
-            //producer.close();
-            try {
-                //Wait for the last packet to get acked
-                retFuture.get();
-            } catch (InterruptedException | ExecutionException e ) {
-                e.printStackTrace();
-            }
+            // producer.close();
+ 
+            //Wait for the last packet to get acked
+            retFuture.get();
         }
 
-        @Override
-        public void run() {
+      @Override
+      public Void call() throws Exception  {
```

Comment:
Just because an interface allows Exception don't mean you can't throw a subclass or none at all. 
Please narrow this to what is actually thrown.

Reply:
sorry i dint get this.. "call" method can throw an exception which propagated to main. hence changed it from Runnable to callable and suitable of invokeAll also; so that no need of counterlatch. i have made it simple i  believe

Reply:
sendFunction() does not throw. runLoop throws, but it is throwing Exception and can be narrowed to a more speffic type. (see my comment there). Once that is done. this function can similarly be narrowed.
Because the interface declares that you can throw Exception, it is OK for your implementation to throw just a subtype for instance IOException. 
This makes the code more readable, and may mean clearer exception handling code in the caller.

Reply:
yes, i have done this .. i defined my own checked Exceptions functional interface